src/wp-includes/user.php

```diff
@@ -2129,7 +2129,8 @@ function wp_insert_user( $userdata ) {
 	}
 
 	// Username must not match an existing user email.
-	if ( email_exists( $user_login ) ) {
+	$existing_user_id = email_exists( $user_login );
+	if ( $existing_user_id && ( ! $update || ( isset( $user_id ) && $existing_user_id !== $user_id ) ) ) {
```

Comment:
Is the `isset( $user_id ) && $existing_user_id !== $user_id` needed here? The `isset( $user_id )` doesn't seem necessary as if `$update` is true, it is always set (see [above](https://github.com/WordPress/wordpress-develop/pull/4257/files#diff-ff818ec19f0f94e4175f72bd2cc4cd67c455c0c50e5b651690a1e312c284ba06L2087)). 

Then the `$existing_user_id !== $user_id` means that only a logged-in user would be able to change that for themselves (an admin may not be able to make changes to another user?). As this bug seems to exist only when the username and email address match, perhaps can check that instead? 

On the other hand looking at the username check above it seems `if ( ! $update && email_exists( $user_login ) )` would be sufficient?

Reply:
I believe you're right about `isset( $user_id )` though, that doesn't seem to be necessary here.

Reply:
> $existing_user_id = The user with that email address
> $user_id = The user being updated.

May be missing something but this means that if user A has a username that matches user B's email address, user A cannot be updated? Shouldn't that be the other way round: prevent an user from changing their email address to match another user's username?

Reply:
> this means that if user A has a username that matches user B's email address, user A cannot be updated?

Just ran a test with the PR - that's correct.

> Shouldn't that be the other way round: prevent an user from changing their email address to match another user's username?

I see your point. I wonder, is there ever a situation where one user should have a username that matches another's email address, or vice versa?

What if:
- We only allow creating if:
  - The username doesn't match an existing username or email address.
  - The email doesn't match an existing username or email address.
- We only allow updating if:
  - Context: Updating with a unique username
    - The username doesn't exist
  - Context: Updating with a unique email address
    - The email address doesn't exist
  - Context: Updating a user without changing the username or email address
    - The username exists, but for the same user.
    - The email address exists, but for the same user.
  - Context: Updating a user to make the username and email address match
    - The username exists as an email address, but it's for the same user.
    - The email address exists as a username, but it's for the same user.

At any rate, it may be better to revert the related changesets while we determine what the "right" path forward is.

Reply:
FWIW, the above does seem to be achievable with some additional checks. Here's the results of some tests I have locally:

![image](https://user-images.githubusercontent.com/79332690/226825334-25fb6b6a-1dda-4d8d-92d6-08fc5561ea30.png)

And the checks (in full):

```php
// Already existed: Username must be unique.
if ( ! $update && username_exists( $user_login ) ) {
	return new WP_Error( 'existing_user_login', __( 'Sorry, that username already exists!' ) );
}

// From this PR (isset() removed + variable rename) Username must not match another user's email address.
$existing_user_email_as_login = email_exists( $user_login );
if ( $existing_user_email_as_login && ( ! $update || $existing_user_email_as_login !== $user_id ) ) {
	return new WP_Error( 'existing_user_email_as_login', __( 'Sorry, that username is not available.' ) );
}

// New: Email address must not match another user's username.
$existing_user_login_as_email = username_exists( $userdata['user_email'] );
if ( $existing_user_login_as_email && ( ! $update || $existing_user_login_as_email !== $user_id ) ) {
	return new WP_Error( 'existing_user_login_as_email', __( 'Sorry, that email address is not available.' ) );
}
```

Reply:
> We only allow creating if:
>   - The username doesn't match an existing username or email address.
>   - The email doesn't match an existing username or email address.

Right. These are the correct requirements imho.

> I wonder, is there ever a situation where one user should have a username that matches another's email address, or vice versa?

There shouldn't be any of these. However WP has no checks to catch these cases. Thinking there should be back-compat to keep updates of existing users with "wrong" username or email still working. This makes it somewhat complicated but seems `$old_user_data` can be useful for that.

Reply:
> We only allow updating if:
> ...

Using an email address as a username is somewhat complex in WP as usernames are generally public. Then changing the email would mean that user now has two emails to test against.

The logic here is good imho, just couple of additions:

- Context: Updating with a unique username (generally usernames cannot be updated, needs a plugin):
  - The username doesn't exist.
  - If the username looks like an email address, the email is not used by another user.
- Context: Updating with a unique email address
  - The email address doesn't exist.
  - The email address is not another user's username.
- Context: Updating a user without changing the username or email address
  - The username exists, but for the same user.
  - The email address exists, but for the same user.
  - Allow the user's email to be another user's username and/or the user's username to be another user's email (back-compat).
- Context: Updating a user to make the username and email address match
  - The username exists as an email address, but it's for the same user.
  - The email address exists as a username, but it's for the same user.
  - Do not allow (obviously) if someone else has the requested username or email. See the back-compat above.

Reply:
Looking a bit more, perhaps this can be implemented as a "long list of `if`s" :) 
Something like:
```
if ( $update ) {
    // Check updated email
    if ( $old_user_data->data->user_email !== $userdata['user_email'] ) {
        if ( [another user's email] ) {
            // Throw error
        }

        if ( $userdata['user_email'] === [another user's username] ) {
            // Throw error
        }
    }

    // Check updated username
    if ( $old_user_data->data->user_login !== $userdata['user_login'] ) {
    ...
}
```